File: Array/Two Pointer Approach/python/Leetcode125.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        """
        **Valid Palindrome**
        **Two Pointer Approach**
        - left & right both starts from 0th & last index of string
        - iterate till left <= right
        - if string[left] != string[right], then return False
        - else, move both pointers (left & right) 1 step forward & backward respectively
        - Finally, return True
        """

        
        st = ""
        for i in s:
            if not i.isalnum():
                continue
            st += i.lower()
            
        left, right = 0, len(st) - 1
        s = st
        while left <= right:
            if s[left] != s[right]:
                return False
            left += 1
            right -= 1
        return True

def main():
    s = "A man, a plan, a canal: Panama"
    solution = Solution()
    result = solution.isPalindrome(s)
    print(result)


if __name__ == "__main__":
    main()

File: Array/Two Pointer Approach/java/Leetcode125.java
class Leetcode125 {
    static boolean isPalindrome(String s) {
        if (s == null) return false;
        if (s.isEmpty()) return true;

        StringBuilder concat_string = new StringBuilder();
        for (char c : s.toCharArray()) {
            if (Character.isLetterOrDigit(c)) {
                concat_string.append(Character.toLowerCase(c));
            }
        }

        String filtered = concat_string.toString();
        System.out.println(filtered);

        int left = 0, right = filtered.length() - 1;
        while (left <= right) {
            if (filtered.charAt(left) != filtered.charAt(right)) {
                return false;
            }
            left++;
            right--;
        }
        return true;
    }


    public static void main(String[] args) {
        boolean flag = isPalindrome("A man, a plan, a canal: Panama");
        System.out.println(flag);
    }
}